1단계 - 엔티티 매핑 과제 PR 요청 드립니다!

taewonh opened the pull request:
안녕하세요!
이번 JPA 과제에 처음으로 참여하게 되었습니다!
앞으로 잘 부탁드리겠습니다!

taewonh committed:
docs(README): 1단계 요구사항 작성

taewonh committed:
chore: DB 설정 정보 추가 및 변경

- H2 Database MySQL 모드 설정
- Hibernate dialect 지정 및 storage engine 설정 innodb로 변경
- SQL 로깅 설정
- 서버 기동 시 테이블 새로 생성하도록 ddl-auto=create 설정

taewonh committed:
feat: Entity 어노테이션 지정

- 요구사항 (객체 - 테이블) 매핑 반영
  - MySQL DDL 예제를 유추하여 동일한 create sql 문을 생성하도록 설정
  - 모든 Entity 상단의 @Entity 지정
  - PK ID 생성 전략 strategy = GenerationType.IDENTITY 지정
    - MySQL 기본 전략인 AUTO_INCREMENT 로 사용
  - 모든 컬럼에 대해 명시적으로 nullable 지정
  - create_at, created_at 의 경우 해당 Entity 정보가 DB에 삽입될 당시에만 입력되어야 하므로 updatable = true 지정

taewonh committed:
refactor: Entity용 공통 추상 클래스 분리

- @Id 만을 가지는 BaseEntity 분리
- createdAt, updatedAt 만을 가지는 DateEntity 분리

taewonh committed:
test: 기본 테스트 정상화

- BaseEntity 분리 과정에서 기존 테스트들이 사용하는 getId 메소드 설정이 protected 되어 테스트들이 사용하는 getId를 Override 해서 get 하도록 설정

taewonh committed:
refactor: repository 패키지 분리

- domain 패키지 내부의 repostory 클래스들을 별도의 패키지로 분리

taewonh committed:
refactor(User): userId Unique 제약 조건 지정

taewonh committed:
test: RepositoryTest 코드 추가

- User, Question, Answer, DeleteHistory Repository 기본 테스트 코드 작성
- MySQL5 Mode 설정 제거
  - RepositoryTest 에서 사용하는 @DataJpaTest 의 경우 Embedded DB 모드만 사용하기 때문에 MySQL5 관련 설정과 충돌이 일어나 해당 설정을 제거 처리함
  - AutoConfigureTestDatabase.Replace.NONE 설정을 사용하면 Embedded DB를 사용하지 않고 datasource 설정을 그대로 사용하지만 추가 요구사항 도출 전까진 해당 옵션을 사용하지 않도록 처리

taewonh committed:
docs(README): 1단계 요구사항 반영 내역 기재

chae-yh commented on src/docs/README.md:
```diff
@@ -0,0 +1,12 @@
+
+## 1단계 요구사항
```
기능 요구사항 잘 정리해주셨네요 👍

chae-yh commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,17 +1,31 @@
+    @Column(nullable = true)
     private Long questionId;
+
+    @Lob
```
적절한 데이터형을 맞춰 주었네요 👍

chae-yh commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -1,17 +1,31 @@
+    @Column(nullable = false)
     private boolean deleted = false;
 
+    protected Answer() { }
```
기본생성자 👍

chae-yh commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -40,14 +54,6 @@ public void toQuestion(Question question) {
-        this.id = id;
-    }
-
     public Long getWriterId() {
```
사용하지 않는 getter / setter는 지워도 좋을 것 같습니다..!

taewonh replied:
넵넵 과제 처음 받고 기본으로 들어가있던 것들이긴 한데 우선 사용하지 않는 메소드들은 모두 제거처리하였습니다!

chae-yh commented on src/main/java/qna/domain/DateEntity.java:
```diff
@@ -0,0 +1,15 @@
+import javax.persistence.Column;
+import javax.persistence.MappedSuperclass;
+
+@MappedSuperclass
```
@MappedSuperclass 사용 👍

chae-yh commented on src/main/java/qna/domain/DateEntity.java:
```diff
@@ -0,0 +1,15 @@
+public abstract class DateEntity extends BaseEntity {
+
+    @Column(nullable = false, updatable = false)
+    protected final LocalDateTime createdAt = LocalDateTime.now();
```
이렇게 값을 넣어줄 수도 있겠지만, jpa auditing 기능을 사용해봐도 좋을 것 같습니다...!

taewonh replied:
@CreatedDate, @LastModifiedDate 적용했습니다!

chae-yh commented on src/main/java/qna/domain/BaseEntity.java:
```diff
@@ -0,0 +1,22 @@
+import javax.persistence.MappedSuperclass;
+
+@MappedSuperclass
+public abstract class BaseEntity {
```
이 구조는 재밌네요 ㅎㅎ 
해당 클래스는 id만 있는데, 각각 필요한 entity가 상속 등의 방법을 피하고, id를 가지는 것과 비교해서 어떤 장점이 있을까요...?

taewonh replied:
앗 사실 그렇게 특별한 의도를 가지고 나눈건 아니었고 Date 만 별도로 모아놓는 DateEntity를 만들면서 같이 나눴던 클래스였는데 크게 의미가 없어보이는 코드가 되었군요.. 이후에 뭔가 다시 분리해야할 상황이 오면 다시 한번 고민해보도록 하겠습니다!

chae-yh commented on src/test/java/qna/repository/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,76 @@
+
+@DataJpaTest
+@DisplayName("Answer")
+public class AnswerRepositoryTest {
```
jpa 의 슬라이싱 테스트 꼼꼼하게 잘 작성되어 있네요 👍

chae-yh commented on src/test/java/qna/repository/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,76 @@
+        Answer saved = repository.save(AnswerTest.A1);
+        Optional<Answer> fetched = repository.findById(saved.getId());
+        assertThat(fetched).isNotEmpty();
+        assertThat(fetched.get()).isEqualTo(saved);
```
사실 현재 answer에는 equals 와 hashCode가 재정의 되어 있지 않기때문에 주소값이 같아야 동일한 객체로 인식합니다. 현재 구조에서는 db에 적재되지 않은, 영속성 컨텍스트에 저장된 값을 가져왔기 때문에 주소 값까지 동일한 객체로 인식하는 상태로 보입니다. 실제로 해당 테스트 수행해보시면 insert 쿼리는 나오는데 select 쿼리가 나오지 않는 것으로 실제 db에 접근하지 않고 캐시된 값을 가져오는 것을 확인하실 수 있어요.

실제 db에 저장하고, 해당 데이터를 가져와서 비교해서 동등성을 확인하는 테스트도 만들어보면 좋을 것 같아요...!

taewonh replied:
넵넵 사실 그부분때문에 고민이 있긴 했는데 일단 객체간 테스트 하는 부분은 모두 없애고, TestEntityManager를 추가하고 데이터를 저장할 때 영속성 컨텍스트를 모두 클리어한 이후 객체 내부 값을 통해 비교하는 방식으로 테스트 코드를 수정했습니다.

이렇게 하니 디버깅을 통해 확인했을 때 서로 다른 객체들이 나와서 동등성은 충분히 검증된다고 생각하는데 혹시 이 방법이 괜찮은 방법일까요??

chae-yh commented on src/test/java/qna/repository/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,76 @@
+    @Test
+    @DisplayName("저장")
+    public void save() {
+        Answer target = AnswerTest.A1;
```
A1 은 테스트를 위해 도메인 객체를 만드는 부분인데 AnswerTest에 두기보다는 AnswerRepositoryTest쪽으로 옮기는 것이 좋지 않을까요...?

taewonh replied:
넵넵 현재 Repository 테스트 모두 기존에 선언되어 있던 도메인 테스트 영역의 static 클래스들을 사용하고 있었는데 모두다 개별적으로 선언해서 사용하는 방식으로 변경했습니다.

chae-yh reviewed (changes requested):
태원님 안녕하세요. jpa 과제를 함께하게 된 채영훈이라고 합니다. 반갑습니다...!
전체적으로, 기능 요구사항을 잘 충족하게 코드 작성주셨습니다 👍 바로 merge 해도 될 것 같지만, 조금 고민해주시면 좋을만한 내용이 있어 코멘트 드리니 확인 부탁드립니다.

감사합니다 : )

taewonh committed:
refactor: 사용되지 않는 getter/setter 제거

- 프로젝트에 기본으로 입력되어 있던 메소드들로 우선적으로 사용되지 않는 메소드들에 한하여 제거 처리

taewonh committed:
fix: test 코드 정상화

- 잘못 제거된 getter로 인해 테스트가 동작하지 않는 문제 수정

taewonh committed:
refactor(BaseEntity): Jpa Auditing 적용

taewonh committed:
refactor: @Entity ID 만을 보유하는 BaseEntity abstract class 제거

- 해당 클래스에서 보유중인 Id 관련 프로퍼티 및 메소드는 Entity 클래스에서 직접 선언하여 사용하도록 변경

taewonh committed:
test: RepositoryTest 동등성 테스트 적용

- save 후 테스트시 동일한 persist context에서 객체를 가져오지 않아야 동등성 테스트가 가능하기 때문에 TestEntityManager를 이용하여 save 후 context를 clear 하는 내부 메소드 추가
- 일부 테스트 메소드 개선

taewonh committed:
test: Repository에서 사용하는 sample entity 객체를 별도로 선언하여 사용하도록 변경

- 기존 도메인 테스트 영역의 static 객체를 사용하지 않고 각 RepositoryTest에서 별도로 선언하여 사용하도록 변경

chae-yh commented on src/test/java/qna/repository/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,102 @@
+    @Autowired
+    private TestEntityManager testEntityManager;
+
+    private static Answer A1;
```
혹시 여기에 static은 왜 필요할까요..? 그리고 A1, D 와 같은 축약어보다는 특징을 나타낼 수 있는 변수명 작성을 해줘도 좋을 것 같습니다 ㅎㅎ

chae-yh commented on src/test/java/qna/repository/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,102 @@
+        Optional<Answer> optional = repository.findById(saved.getId());
+        assertThat(optional).isNotEmpty();
+        Answer fetched = optional.get();
+        assertThat(fetched.getId()).isEqualTo(saved.getId());
```
id 비교 좋네요 👍
동등성 비교는 isEqualTo를 통해 두 객체를 직접 비교해주고, 비교 조건을 equlas와 hashCode 재정의 통해 명시해주는데 해당 부분은 인지하고 계신 것 같아서 우선 이정도로 해도 될 것 같습니다...!

chae-yh commented on src/main/java/qna/domain/DateEntity.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.data.jpa.domain.support.AuditingEntityListener;
+
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
```
피드백 반영 👍

chae-yh commented on src/test/java/qna/repository/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,102 @@
+    private AnswerRepository repository;
+
+    @Autowired
+    private TestEntityManager testEntityManager;
```
보통은 EntityManger를 그냥 쓰곤했는데, TestEntityManger도 있군요 ㅎㅎ 
하나 배워갑니다 : )

chae-yh commented on src/test/java/qna/repository/AnswerRepositoryTest.java:
```diff
@@ -0,0 +1,102 @@
+        assertThat(optional).isEmpty();
+    }
+
+    private Answer saveAndRefetch(Answer answer) {
```
테스트 클래스라 크게 상관은 없지만, 각 메서드는 가능한 한가지의 역할만 해주도록 하는 것이 좋을 것 같아요...!

chae-yh reviewed (approved):
태원님 안녕하세요. 피드백 드린 내용 잘 반영해주셨습니다 👍 
몇가지 궁금한 점이 있어 코멘트 드렸으나, 소소한 내용으로 살짝 확인만 하시면 될 것 같아요

고생하셨고, 다음 스텝 진행부탁드립니다 : )
